lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -2193,12 +2197,116 @@ constructTransaction ctx genChange (ApiT wid) body = do
+        , GenChange s
+        , Bounded (Index (AddressIndexDerivationType k) 'AddressK)
+        )
+    => ctx
+    -> ArgGenChange s
     -> ApiT WalletId
     -> ApiBalanceTransactionPostData n
     -> Handler (ApiConstructTransaction n)
-balanceTransaction _ctx (ApiT _wid) _body =
-    liftHandler $ throwE ErrBalanceTxNotImplemented
+balanceTransaction ctx genChange (ApiT wid) body = do
+    pp <- liftIO $ NW.currentProtocolParameters nl
+
+    --TODO  deal with coll and validity and delegations
+    let (Tx _id _fee _coll _inps outs wdrlMap mdM _validity) = txIncoming
+
+    ((FeeEstimation feeMin _), unsignedtx) <- withWorkerCtx ctx wid liftE liftE $ \wrk -> do
+
+        (acct, _, _) <- liftHandler $ W.readRewardAccount @_ @s @k @n wrk wid
+        (wdrl, _) <-
+            if Map.member acct wdrlMap then
+                mkRewardAccountBuilder @_ @s @_ @n ctx wid (Just SelfWithdrawal)
+            else
+                mkRewardAccountBuilder @_ @s @_ @n ctx wid Nothing
+        let executionFee = calcScriptExecutionCost tl pp sealedTxIncoming
+        ttl <- liftIO $ W.getTxExpiry ti Nothing
+        let txCtx = defaultTransactionCtx
+                { txPlutusScriptExecutionCost = executionFee
+                , txMetadata = mdM
+                , txWithdrawal = wdrl
+                , txTimeToLive = ttl
+                }
+
+        (internalUtxoAvailable, wallet, pendingTxs) <-
+            liftHandler $ W.readWalletUTxOIndex @_ @s @k wrk wid
+
+        let externalSelectedUtxo =
+                UTxOIndex.fromSequence externalInputs
+
+        let utxoAvailable = UTxOSelection.fromIndexPair
+                (internalUtxoAvailable, externalSelectedUtxo)
+
+        let runSelection = W.selectAssets @_ @s @k wrk W.SelectAssetsParams
+                { outputs = outs
+                , pendingTxs
+                , txContext = txCtx
+                , utxoAvailableForInputs = utxoAvailable
+                , utxoAvailableForCollateral =
+                    UTxOIndex.toUTxO internalUtxoAvailable
+                , wallet
+                } getFee
+              where getFee = const (selectionDelta TokenBundle.getCoin)
+
+        let transform s sel' =
+                W.assignChangeAddresses genChange sel' s
+                & uncurry (W.selectionToUnsignedTx (txWithdrawal txCtx))
+        unsignedtx <- liftHandler $ W.selectAssets @_ @s @k wrk W.SelectAssetsParams
+            { outputs = outs
+            , pendingTxs
+            , txContext = txCtx
+            , utxoAvailableForInputs = utxoAvailable
+            , utxoAvailableForCollateral =
+                    UTxOIndex.toUTxO internalUtxoAvailable
+            , wallet
+            }
+            transform
+
+        fee <- liftHandler $ W.estimateFee runSelection
+        return (fee, unsignedtx)
+
+    let UnsignedTx colls inps _outs change _wdrl = unsignedtx
+    let extraBody = ExtraTxBodyContent
+            { extraInputs = (\(txin, _, _) -> txin) <$> F.toList inps
```

Comment:
Won't `inps` already include the pre-selected inputs from `sealedTxIncoming`?

`updateTx` would then try to append the whole `inps` to the pre-selected inputs.

Although as the ledger represents the inputs as a `Set`, it might not actually matter in the end.

Reply:
yes, we may filter out preselected inputs. although ledger stores inputs as set so it should not affect txbody or fee.